In [9]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as dt

pd.set_option('display.float_format', lambda x: '%.2f' % x)

# Online Retail II Dataset

- InvoiceNo: Invoice number. Nominal. A 6-digit integral number uniquely assigned to each transaction. If this code starts with the letter 'c', it indicates a cancellation.

- StockCode: Product (item) code. Nominal. A 5-digit integral number uniquely assigned to each distinct product.

- Description: Product (item) name. Nominal.

- Quantity: The quantities of each product (item) per transaction. Numeric.

- InvoiceDate: Invice date and time. Numeric. The day and time when a transaction was generated.

- UnitPrice: Unit price. Numeric. Product price per unit in sterling (£).

- CustomerID: Customer number. Nominal. A 5-digit integral number uniquely assigned to each customer.

- Country: Country name. Nominal. The name of the country where a customer resides.

In [10]:
dataset = pd.read_csv('online_retail_II.csv', dtype={'Customer ID':float, 'Invoice':str})

In [11]:
dataset = pd.read_csv('online_retail_II.csv', dtype={'Customer ID':float, 'Invoice':str})

In [12]:
dataset.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

           InvoiceDate  Price  Customer ID         Country  
0  2009-12-01 07:45:00   6.95     13085.00  United Kingdom  
1  2009-12-01 07:45:00   6.75     13085.00  United Kingdom  
2  2009-12-01 07:45:00   6.75     13085.00  United Kingdom  
3  2009-12-01 07:45:00   2.10     13085.00  United Kingdom  
4  2009-12-01 07:45:00   1.25     13085.00  United Kingdom

In [13]:
dataset.shape

(1067371, 8)

# Information of Data

In [14]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  object 
 1   StockCode    1067371 non-null  object 
 2   Description  1062989 non-null  object 
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  object 
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 65.1+ MB


In [15]:
dataset.describe()

Quantity      Price  Customer ID
count 1067371.00 1067371.00    824364.00
mean        9.94       4.65     15324.64
std       172.71     123.55      1697.46
min    -80995.00  -53594.36     12346.00
25%         1.00       1.25     13975.00
50%         3.00       2.10     15255.00
75%        10.00       4.15     16797.00
max     80995.00   38970.00     18287.00

**We have a negative values in Quantity and Price**

In [46]:
dataset[(dataset['Quantity'] < 0) | (dataset['Price'] < 0)]

Invoice StockCode                       Description  Quantity  \
178      C489449     22087          PAPER BUNTING WHITE LACE       -12   
179      C489449    85206A      CREAM FELT EASTER EGG BASKET        -6   
180      C489449     21895     POTTING SHED SOW 'N' GROW SET        -4   
181      C489449     21896                POTTING SHED TWINE        -6   
182      C489449     22083        PAPER CHAIN KIT RETRO SPOT       -12   
...          ...       ...                               ...       ...   
1065910  C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
1067002  C581499         M                            Manual        -1   
1067176  C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
1067177  C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
1067178  C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

                 InvoiceDate   Price  Customer ID         Country  
178      2009-12-01 10:33:00    2.95      16321.0       Australia  
179      2009-12-01 10:33:00    1.65      16321.0       Australia  
180      2009-12-01 10:33:00    4.25      16321.0       Australia  
181      2009-12-01 10:33:00    2.10      16321.0       Australia  
182      2009-12-01 10:33:00    2.95      16321.0       Australia  
...                      ...     ...          ...             ...  
1065910  2011-12-09 09:57:00    0.83      14397.0  United Kingdom  
1067002  2011-12-09 10:28:00  224.69      15498.0  United Kingdom  
1067176  2011-12-09 11:57:00   10.95      15311.0  United Kingdom  
1067177  2011-12-09 11:58:00    1.25      17315.0  United Kingdom  
1067178  2011-12-09 11:58:00    1.25      17315.0  United Kingdom  

[22955 rows x 8 columns]

In [47]:
dataset[(dataset['Quantity'] < 0) | (dataset['Price'] < 0)].describe()

Quantity         Price   Customer ID
count  22955.000000  22955.000000  18744.000000
mean     -46.354738     31.177929  15014.418694
std      788.164950    767.464100   1698.413277
min   -80995.000000 -53594.360000  12346.000000
25%      -11.000000      0.850000  13552.000000
50%       -2.000000      2.250000  14911.000000
75%       -1.000000      4.950000  16401.000000
max        1.000000  38970.000000  18287.000000

## A. Data Cleaning and Preparation

<span style='color:Indianred'>**Duplicated Values**</span>

In [48]:
dupli = dataset[dataset.duplicated(subset=dataset.columns,keep=False)]
dupli.shape

(67242, 8)

In [49]:
dupli

Invoice StockCode                        Description  Quantity  \
362      489517     21913     VINTAGE SEASIDE JIGSAW PUZZLES         1   
363      489517     21912           VINTAGE SNAKES & LADDERS         1   
365      489517     21821   GLITTER STAR GARLAND WITH BELLS          1   
367      489517     22319  HAIRCLIPS FORTIES FABRIC ASSORTED        12   
368      489517     22130   PARTY CONE CHRISTMAS DECORATION          6   
...         ...       ...                                ...       ...   
1067136  581538     22068        BLACK PIRATE TREASURE CHEST         1   
1067150  581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
1067153  581538     22992             REVOLVER WOODEN RULER          1   
1067160  581538     22694                       WICKER STAR          1   
1067162  581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

                 InvoiceDate  Price  Customer ID         Country  
362      2009-12-01 11:34:00   3.75      16329.0  United Kingdom  
363      2009-12-01 11:34:00   3.75      16329.0  United Kingdom  
365      2009-12-01 11:34:00   3.75      16329.0  United Kingdom  
367      2009-12-01 11:34:00   0.65      16329.0  United Kingdom  
368      2009-12-01 11:34:00   0.85      16329.0  United Kingdom  
...                      ...    ...          ...             ...  
1067136  2011-12-09 11:34:00   0.39      14446.0  United Kingdom  
1067150  2011-12-09 11:34:00   2.49      14446.0  United Kingdom  
1067153  2011-12-09 11:34:00   1.95      14446.0  United Kingdom  
1067160  2011-12-09 11:34:00   2.10      14446.0  United Kingdom  
1067162  2011-12-09 11:34:00   2.08      14446.0  United Kingdom  

[67242 rows x 8 columns]

**We've got a 67242 rows duplicated values, so we're gonna removing all duplicated and keep first values from them.**

In [50]:
df = dataset.copy()

In [51]:
df.drop_duplicates(keep='first', inplace=True)

In [52]:
df.shape

(1033036, 8)

In [53]:
print('The data was decreasing ',dataset.shape[0]-df.shape[0], ' rows, after a removing a duplicates and keep first values from them')

The data was decreasing  34335  rows, after a removing a duplicates and keep first values from them


<span style='color:Indianred'>**Checking Missing Values**</span>

In [39]:
pd.DataFrame(round(df.isnull().sum()/df.shape[0] * 100,3), columns = ['Missing (%)'])

Missing (%)
Invoice            0.000
StockCode          0.000
Description        0.417
Quantity           0.000
InvoiceDate        0.000
Price              0.000
Customer ID       22.728
Country            0.000

**Imputing a Missing values in a Customer ID columns can't take a mean/mode/median as a methods, it just because we never know what kind of customers or where the place customers has purchase the product. We gonna keep that missing values for further analysis**

In [40]:
miss_val = dataset[(dataset['Customer ID'].isna()) | (dataset['Description'].isna())]
miss_val

Invoice StockCode                     Description  Quantity  \
263      489464     21733                    85123a mixed       -96   
283      489463     71477                           short      -240   
284      489467    85123A                     21733 mixed      -192   
470      489521     21646                             NaN       -50   
577      489525    85226C       BLUE PULL BACK RACING CAR         1   
...         ...       ...                             ...       ...   
1066997  581498    85099B         JUMBO BAG RED RETROSPOT         5   
1066998  581498    85099C  JUMBO  BAG BAROQUE BLACK WHITE         4   
1066999  581498     85150   LADIES & GENTLEMEN METAL SIGN         1   
1067000  581498     85174               S/4 CACTI CANDLES         1   
1067001  581498       DOT                  DOTCOM POSTAGE         1   

                 InvoiceDate    Price  Customer ID         Country  
263      2009-12-01 10:52:00     0.00          NaN  United Kingdom  
283      2009-12-01 10:52:00     0.00          NaN  United Kingdom  
284      2009-12-01 10:53:00     0.00          NaN  United Kingdom  
470      2009-12-01 11:44:00     0.00          NaN  United Kingdom  
577      2009-12-01 11:49:00     0.55          NaN  United Kingdom  
...                      ...      ...          ...             ...  
1066997  2011-12-09 10:26:00     4.13          NaN  United Kingdom  
1066998  2011-12-09 10:26:00     4.13          NaN  United Kingdom  
1066999  2011-12-09 10:26:00     4.96          NaN  United Kingdom  
1067000  2011-12-09 10:26:00    10.79          NaN  United Kingdom  
1067001  2011-12-09 10:26:00  1714.17          NaN  United Kingdom  

[243007 rows x 8 columns]

### Understanding Features

### a. `Invoice`

- InvoiceNo: Invoice number. Nominal. A 6-digit integral number uniquely assigned to each transaction. If this code starts with the letter 'c', it indicates a cancellation.

variabel `nc`  =  `no_cancel` dataset which those data doesn't have a `cancel_condition`

In [17]:
cancel_condition = df['Invoice'].apply(lambda x: x[0] == 'C')
nc = df[~cancel_condition]
display(nc, df[cancel_condition])

Invoice StockCode                          Description  Quantity  \
0        489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1        489434    79323P                   PINK CHERRY LIGHTS        12   
2        489434    79323W                  WHITE CHERRY LIGHTS        12   
3        489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4        489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...         ...       ...                                  ...       ...   
1067366  581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
1067367  581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
1067368  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
1067369  581587     22138        BAKING SET 9 PIECE RETROSPOT          3   
1067370  581587      POST                              POSTAGE         1   

                 InvoiceDate  Price  Customer ID         Country  
0        2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1        2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2        2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3        2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4        2009-12-01 07:45:00   1.25      13085.0  United Kingdom  
...                      ...    ...          ...             ...  
1067366  2011-12-09 12:50:00   2.10      12680.0          France  
1067367  2011-12-09 12:50:00   4.15      12680.0          France  
1067368  2011-12-09 12:50:00   4.15      12680.0          France  
1067369  2011-12-09 12:50:00   4.95      12680.0          France  
1067370  2011-12-09 12:50:00  18.00      12680.0          France  

[1013932 rows x 8 columns]

Invoice StockCode                       Description  Quantity  \
178      C489449     22087          PAPER BUNTING WHITE LACE       -12   
179      C489449    85206A      CREAM FELT EASTER EGG BASKET        -6   
180      C489449     21895     POTTING SHED SOW 'N' GROW SET        -4   
181      C489449     21896                POTTING SHED TWINE        -6   
182      C489449     22083        PAPER CHAIN KIT RETRO SPOT       -12   
...          ...       ...                               ...       ...   
1065910  C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
1067002  C581499         M                            Manual        -1   
1067176  C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
1067177  C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
1067178  C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

                 InvoiceDate   Price  Customer ID         Country  
178      2009-12-01 10:33:00    2.95      16321.0       Australia  
179      2009-12-01 10:33:00    1.65      16321.0       Australia  
180      2009-12-01 10:33:00    4.25      16321.0       Australia  
181      2009-12-01 10:33:00    2.10      16321.0       Australia  
182      2009-12-01 10:33:00    2.95      16321.0       Australia  
...                      ...     ...          ...             ...  
1065910  2011-12-09 09:57:00    0.83      14397.0  United Kingdom  
1067002  2011-12-09 10:28:00  224.69      15498.0  United Kingdom  
1067176  2011-12-09 11:57:00   10.95      15311.0  United Kingdom  
1067177  2011-12-09 11:58:00    1.25      17315.0  United Kingdom  
1067178  2011-12-09 11:58:00    1.25      17315.0  United Kingdom  

[19104 rows x 8 columns]

The invoice is a unique identifier, assuming that numbers denote the customers purchasing with a multiple items. So we have to check.

In [18]:
nc[nc['Invoice'] == '581587']

Invoice StockCode                      Description  Quantity  \
1067355  581587     22631         CIRCUS PARADE LUNCH BOX         12   
1067356  581587     22556   PLASTERS IN TIN CIRCUS PARADE         12   
1067357  581587     22555        PLASTERS IN TIN STRONGMAN        12   
1067358  581587     22728        ALARM CLOCK BAKELIKE PINK         4   
1067359  581587     22727        ALARM CLOCK BAKELIKE RED          4   
1067360  581587     22726       ALARM CLOCK BAKELIKE GREEN         4   
1067361  581587     22730       ALARM CLOCK BAKELIKE IVORY         4   
1067362  581587     22367  CHILDRENS APRON SPACEBOY DESIGN         8   
1067363  581587     22629              SPACEBOY LUNCH BOX         12   
1067364  581587     23256      CHILDRENS CUTLERY SPACEBOY          4   
1067365  581587     22613      PACK OF 20 SPACEBOY NAPKINS        12   
1067366  581587     22899     CHILDREN'S APRON DOLLY GIRL          6   
1067367  581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
1067368  581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
1067369  581587     22138    BAKING SET 9 PIECE RETROSPOT          3   
1067370  581587      POST                          POSTAGE         1   

                 InvoiceDate  Price  Customer ID Country  
1067355  2011-12-09 12:50:00   1.95      12680.0  France  
1067356  2011-12-09 12:50:00   1.65      12680.0  France  
1067357  2011-12-09 12:50:00   1.65      12680.0  France  
1067358  2011-12-09 12:50:00   3.75      12680.0  France  
1067359  2011-12-09 12:50:00   3.75      12680.0  France  
1067360  2011-12-09 12:50:00   3.75      12680.0  France  
1067361  2011-12-09 12:50:00   3.75      12680.0  France  
1067362  2011-12-09 12:50:00   1.95      12680.0  France  
1067363  2011-12-09 12:50:00   1.95      12680.0  France  
1067364  2011-12-09 12:50:00   4.15      12680.0  France  
1067365  2011-12-09 12:50:00   0.85      12680.0  France  
1067366  2011-12-09 12:50:00   2.10      12680.0  France  
1067367  2011-12-09 12:50:00   4.15      12680.0  France  
1067368  2011-12-09 12:50:00   4.15      12680.0  France  
1067369  2011-12-09 12:50:00   4.95      12680.0  France  
1067370  2011-12-09 12:50:00  18.00      12680.0  France

In [19]:
nc.groupby('Customer ID').count()

Invoice  StockCode  Description  Quantity  InvoiceDate  Price  \
Customer ID                                                                  
12346.0           34         34           34        34           34     34   
12347.0          222        222          222       222          222    222   
12348.0           51         51           51        51           51     51   
12349.0          175        175          175       175          175    175   
12350.0           17         17           17        17           17     17   
...              ...        ...          ...       ...          ...    ...   
18283.0          938        938          938       938          938    938   
18284.0           28         28           28        28           28     28   
18285.0           12         12           12        12           12     12   
18286.0           67         67           67        67           67     67   
18287.0          155        155          155       155          155    155   

             Country  
Customer ID           
12346.0           34  
12347.0          222  
12348.0           51  
12349.0          175  
12350.0           17  
...              ...  
18283.0          938  
18284.0           28  
18285.0           12  
18286.0           67  
18287.0          155  

[5881 rows x 7 columns]

I think that customers has purchased multiple items, so it's makes sense. Doesn't need a manipulation in this case.

**From understanding `Invoice` feature, separate the *cancelation* purchase is the best way to continue analyzing this data. Holding that *Cancelation* or *C* at this data make a bias for further analysis so I'm gonna remove that.**

### b. `StockCode`

- StockCode: Product (item) code. Nominal. A 5-digit integral number uniquely assigned to each distinct product.

**From `StockCode` *feature* description that item code have a 5 digit number uniquely**

In [21]:
a = '79323W'
a[:5].isnumeric()

True

In [22]:
stockrules = nc['StockCode'].apply(lambda x: x[:5].isnumeric())
nc[stockrules]

Invoice StockCode                          Description  Quantity  \
0        489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1        489434    79323P                   PINK CHERRY LIGHTS        12   
2        489434    79323W                  WHITE CHERRY LIGHTS        12   
3        489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4        489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...         ...       ...                                  ...       ...   
1067365  581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
1067366  581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
1067367  581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
1067368  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
1067369  581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

                 InvoiceDate  Price  Customer ID         Country  
0        2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1        2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2        2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3        2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4        2009-12-01 07:45:00   1.25      13085.0  United Kingdom  
...                      ...    ...          ...             ...  
1067365  2011-12-09 12:50:00   0.85      12680.0          France  
1067366  2011-12-09 12:50:00   2.10      12680.0          France  
1067367  2011-12-09 12:50:00   4.15      12680.0          France  
1067368  2011-12-09 12:50:00   4.15      12680.0          France  
1067369  2011-12-09 12:50:00   4.95      12680.0          France  

[1009142 rows x 8 columns]

**According thr description of `StockCode` that stock code have a 5 digit number. Assuming additional last words are variances of Products**

In [23]:
print(f"So we have {nc[~stockrules].shape[0]} rows doesn't meet the condition of stock rules")

So we have 4790 rows doesn't meet the condition of stock rules


In [24]:
ds = nc[~stockrules][['StockCode']]
ds['StockCode'].unique()

array(['POST', 'DCGS0058', 'DCGS0068', 'DOT', 'M', 'DCGS0004', 'DCGS0076',
       'C2', 'BANK CHARGES', 'DCGS0003', 'TEST001', 'gift_0001_80',
       'DCGS0072', 'gift_0001_20', 'DCGS0044', 'TEST002', 'gift_0001_10',
       'gift_0001_50', 'DCGS0066N', 'gift_0001_30', 'PADS', 'ADJUST',
       'gift_0001_40', 'gift_0001_60', 'gift_0001_70', 'gift_0001_90',
       'DCGSSGIRL', 'DCGS0006', 'DCGS0016', 'DCGS0027', 'DCGS0036',
       'DCGS0039', 'DCGS0060', 'DCGS0056', 'DCGS0059', 'GIFT', 'DCGSLBOY',
       'm', 'DCGS0053', 'DCGS0062', 'DCGS0037', 'DCGSSBOY', 'DCGSLGIRL',
       'DCGS0069', 'DCGS0070', 'DCGS0075', 'D', 'B', 'S', 'DCGS0041',
       'ADJUST2', 'C3', 'SP1002', 'AMAZONFEE', 'DCGS0055', 'DCGS0074',
       'DCGS0057', 'DCGS0073', 'DCGS0071', 'DCGS0066P', 'DCGS0067'],
      dtype=object)

Fortunately, there is only a small number of data points like this so lets remove them. Does it matter if I am suggesting to a buyer a pink or white item? Yes, it does! Let's say that a buyer is decorating a bedroom with a specific color pallet around pink, that buyer will be more susceptible to a suggested item within his desired color pallet. If enough buyers have a "thematic buying pattern" in the dataset it is possible that any future models build upon this dataset will be able to pick this knowledge. Anyways, the collapsing of data leads to an information loss which I really want to avoid. So, the answer is no, collapsing seems like a bad idea here.

In [25]:
nc = nc[stockrules]

In [26]:
nc.shape

(1009142, 8)

### c. Customer ID

- `Customer ID`: Customer number. Nominal. A 5-digit integral number uniquely assigned to each customer.

In [27]:
no_reg = nc[nc['Customer ID'].isnull()]
reg = nc[~nc['Customer ID'].isnull()]

print(f"{no_reg.shape[0] * 100 / reg.shape[0]}% of the data point have missing 'Customer ID'")

29.937409626376287% of the data point have missing 'Customer ID'


In [28]:
no_reg.shape

(232505, 8)

In [29]:
reg.shape

(776637, 8)

#### Imputing Customer ID

Identifying `Customer ID` from `Invoice` features, the logic are that pruchased item with the same `Invoice` made by the same user. If the missing `Customer ID` are due to unregistered users/removed by request, this approach is not going to work.

In [30]:
invoice_no_reg = no_reg['Invoice'].unique()
invoice_with_reg = reg['Invoice'].unique()

cross_cust_id = [x for x in invoice_no_reg if x in invoice_with_reg]

print(f"There are {len(cross_cust_id)} overlapping entries between those with \
and without 'Customer ID' regarding the 'Invoice'")
cross_cust_id

There are 0 overlapping entries between those with and without 'Customer ID' regarding the 'Invoice'


[]

There is no overlapping `Invoice` unique features from `Null` Values, so the conclusion I'm gonna input *dummies* `Customer ID` which related to buy timing and `Invoice`. Assuming that each `Invoice` corresponds to a single users.

In [31]:
invc_count_cust = reg.groupby(by='Invoice')[['Customer ID']].count()
display(invc_count_cust)
print(f"The total of one single user purchasing one item within {invc_count_cust[invc_count_cust['Customer ID'] == 1].sum()}")
print(f"Counting an Invoice which related by Customer ID on Purchasing {invc_count_cust.shape[0]}")

Customer ID
Invoice             
489434             8
489435             4
489436            19
489437            23
489438            17
...              ...
581583             2
581584             2
581585            21
581586             4
581587            15

[36597 rows x 1 columns]

The total of one single user purchasing one item within Customer ID    2741
dtype: int64
Counting an Invoice which related by Customer ID on Purchasing 36597


In [58]:
nc.reset_index(inplace=True)

In [79]:
nc['Customer ID'].fillna(0, inplace=True)

In [80]:
unique_invoice = no_reg['Invoice'].unique()

mapper = {}
for i in range(len(unique_invoice)):
    mapper[unique_invoice[i]] = f'imp{i}'

# mapper = {v:k for k,v in mapper.items()}
for i in range(len(nc['Customer ID'])):
    if nc.loc[i, 'Customer ID'] == 0 and nc.loc[i, 'Invoice'] in mapper.keys():
        nc.loc[i, 'Customer ID'] = mapper[nc.loc[i, 'Invoice']]

In [81]:
nc[nc['Customer ID'].isna()]

Empty DataFrame
Columns: [index, Invoice, StockCode, Description, Quantity, InvoiceDate, Price, Customer ID, Country]
Index: []

<hr>

Take a long time to impute that missing values, so i'm gonna convert that `DataFrame` to csv again to speed up the process.

In [ ]:
nc.to_csv('Online Retail II Impute NaN Customer.csv')

<hr>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as dt

import warnings 
warnings.filterwarnings("ignore")

In [2]:
nc = pd.read_csv('Online Retail II Impute NaN Customer.csv')

In [3]:
nc.drop(columns=['Unnamed: 0', 'index'], inplace=True)

In [4]:
nc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1009142 entries, 0 to 1009141
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1009142 non-null  int64  
 1   StockCode    1009142 non-null  object 
 2   Description  1004930 non-null  object 
 3   Quantity     1009142 non-null  int64  
 4   InvoiceDate  1009142 non-null  object 
 5   Price        1009142 non-null  float64
 6   Customer ID  1009142 non-null  object 
 7   Country      1009142 non-null  object 
dtypes: float64(1), int64(2), object(5)
memory usage: 61.6+ MB


In [5]:
nc.shape

(1009142, 8)

### d. `Quantity` and `Price`

- Quantity: The quantities of each product (item) per transaction. Numeric.

- UnitPrice: Unit price. Numeric. Product price per unit in sterling (£).

Checking the business logic where the quantity and price is impossible to have a negative values. I'm gonna separate all of these negative values, features *Description* didn't explain what negative values are.

In [6]:
nc[(nc['Price'] < 0) | (nc['Quantity'] < 0)]

Invoice StockCode   Description  Quantity          InvoiceDate  \
238       489464     21733  85123a mixed       -96  2009-12-01 10:52:00   
258       489463     71477         short      -240  2009-12-01 10:52:00   
259       489467    85123A   21733 mixed      -192  2009-12-01 10:53:00   
430       489521     21646           NaN       -50  2009-12-01 11:44:00   
2947      489655     20683           NaN       -44  2009-12-01 17:26:00   
...          ...       ...           ...       ...                  ...   
1002696   581210     23395         check       -26  2011-12-07 18:36:00   
1002698   581212     22578          lost     -1050  2011-12-07 18:38:00   
1002699   581213     22576         check       -30  2011-12-07 18:38:00   
1004270   581226     23090       missing      -338  2011-12-08 09:56:00   
1006206   581422     23169       smashed      -235  2011-12-08 15:24:00   

         Price Customer ID         Country  
238        0.0        imp0  United Kingdom  
258        0.0        imp1  United Kingdom  
259        0.0        imp2  United Kingdom  
430        0.0        imp3  United Kingdom  
2947       0.0       imp25  United Kingdom  
...        ...         ...             ...  
1002696    0.0     imp8099  United Kingdom  
1002698    0.0     imp8101  United Kingdom  
1002699    0.0     imp8102  United Kingdom  
1004270    0.0     imp8105  United Kingdom  
1006206    0.0     imp8111  United Kingdom  

[3362 rows x 8 columns]

In [7]:
nc = nc[~((nc['Price'] < 0) | (nc['Quantity'] < 0))]

In [8]:
nc.shape

(1005780, 8)

### e. `InvoiceDate`

- InvoiceDate: Invice date and time. Numeric. The day and time when a transaction was generated..

From the information of the data, this feature still an `object` so we have to change it to date time format, and change the number of month to Month as locale’s abbreviated name. Splitting up the time gonna be useful enough, hope leads me up to get beautiful insight for this `dataset`.

*Python strftime reference :*  https://strftime.org/

In [9]:
nc['InvoiceDate'].head()

0    2009-12-01 07:45:00
1    2009-12-01 07:45:00
2    2009-12-01 07:45:00
3    2009-12-01 07:45:00
4    2009-12-01 07:45:00
Name: InvoiceDate, dtype: object

In [10]:
nc['InvoiceDate'] = pd.to_datetime(nc['InvoiceDate'], format='%Y-%m-%d %H:%M:%S')

In [11]:
nc['InvoiceDate'].head()

0   2009-12-01 07:45:00
1   2009-12-01 07:45:00
2   2009-12-01 07:45:00
3   2009-12-01 07:45:00
4   2009-12-01 07:45:00
Name: InvoiceDate, dtype: datetime64[ns]

In [12]:
nc['Day of Week'] = nc['InvoiceDate'].dt.strftime('%A')

In [13]:
nc['Date'] = nc['InvoiceDate'].dt.strftime('%d')

In [14]:
nc['Month'] = nc['InvoiceDate'].dt.strftime('%b')

In [15]:
nc['Year'] = nc['InvoiceDate'].dt.strftime('%Y')

In [16]:
nc['Day of Year'] = nc['InvoiceDate'].dt.strftime('%j') 

In [17]:
nc['Week of Year'] = nc['InvoiceDate'].dt.strftime('%W')

In [18]:
nc['Time Hour'] = nc['InvoiceDate'].dt.strftime('%X')

In [19]:
nc['Hour'] = nc['InvoiceDate'].dt.hour

In [20]:
nc.head(2)

Invoice StockCode                          Description  Quantity  \
0   489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1   489434    79323P                   PINK CHERRY LIGHTS        12   

          InvoiceDate  Price Customer ID         Country Day of Week Date  \
0 2009-12-01 07:45:00   6.95     13085.0  United Kingdom     Tuesday   01   
1 2009-12-01 07:45:00   6.75     13085.0  United Kingdom     Tuesday   01   

  Month  Year Day of Year Week of Year Time Hour  Hour  
0   Dec  2009         335           48  07:45:00     7  
1   Dec  2009         335           48  07:45:00     7

### f. `Description`

- Description: Product (item) name. Nominal.

Checking The missing values

In [21]:
display(nc[nc['Description'].isna()])
print(f"We have {nc['Description'].isna().sum()} missing values from Description Rows")

Invoice StockCode Description  Quantity         InvoiceDate  Price  \
2994      489659     21350         NaN       230 2009-12-01 17:39:00    0.0   
3549      489781     84292         NaN        17 2009-12-02 11:45:00    0.0   
6146      489882    35751C         NaN        12 2009-12-02 16:22:00    0.0   
6322      489898    79323G         NaN       954 2009-12-03 09:40:00    0.0   
6348      489903     21166         NaN        48 2009-12-03 09:57:00    0.0   
...          ...       ...         ...       ...                 ...    ...   
1001137   581103     22689         NaN         4 2011-12-07 11:58:00    0.0   
1002689   581203     23406         NaN        15 2011-12-07 18:31:00    0.0   
1002695   581209     21620         NaN         6 2011-12-07 18:35:00    0.0   
1004324   581234     72817         NaN        27 2011-12-08 10:33:00    0.0   
1005857   581408     85175         NaN        20 2011-12-08 14:06:00    0.0   

        Customer ID         Country Day of Week Date Month  Year Day of Year  \
2994          imp26  United Kingdom     Tuesday   01   Dec  2009         335   
3549          imp29  United Kingdom   Wednesday   02   Dec  2009         336   
6146          imp37  United Kingdom   Wednesday   02   Dec  2009         336   
6322          imp39  United Kingdom    Thursday   03   Dec  2009         337   
6348          imp42  United Kingdom    Thursday   03   Dec  2009         337   
...             ...             ...         ...  ...   ...   ...         ...   
1001137     imp8084  United Kingdom   Wednesday   07   Dec  2011         341   
1002689     imp8092  United Kingdom   Wednesday   07   Dec  2011         341   
1002695     imp8098  United Kingdom   Wednesday   07   Dec  2011         341   
1004324     imp8106  United Kingdom    Thursday   08   Dec  2011         342   
1005857     imp8110  United Kingdom    Thursday   08   Dec  2011         342   

        Week of Year Time Hour  Hour  
2994              48  17:39:00    17  
3549              48  11:45:00    11  
6146              48  16:22:00    16  
6322              48  09:40:00     9  
6348              48  09:57:00     9  
...              ...       ...   ...  
1001137           49  11:58:00    11  
1002689           49  18:31:00    18  
1002695           49  18:35:00    18  
1004324           49  10:33:00    10  
1005857           49  14:06:00    14  

[1605 rows x 16 columns]

We have 1605 missing values from Description Rows


#### Cross Validation Check

It's a logical whereas the `StockCode` represent `Description` as a unique Code.

Separating `NaN` values  `DataFrame` and *Unique code `pd.Series`* before cross checking.

In [22]:
# DataFrame
desc_nan = nc[nc['Description'].isna()]
desc_nonan = nc[~nc['Description'].isna()]

# Unique Code

stck_unq = desc_nonan['StockCode'].unique()
stck_unq_nan = desc_nan['StockCode'].unique()

cross_stck = [x for x in stck_unq_nan if x in stck_unq]

print(f"There are {len(cross_stck)} overlapping entries between those with \
and without 'Description' regarding the 'Stock Code'")

There are 986 overlapping entries between those with and without 'Description' regarding the 'Stock Code'


Checking overlapping result into Description

In [23]:
idx_stck = desc_nonan.set_index('StockCode')
check_stck = idx_stck.loc[cross_stck, :][['Description']].drop_duplicates(keep='first').reset_index()
# check_stck.reset_index(inplace=True)
check_stck.drop_duplicates(subset='StockCode', keep='first', inplace=True)
display(check_stck.head(), check_stck.shape)

StockCode                    Description
0     21350           GINGHAM HEART WREATH
1     84292  ROSE PINK METAL FOLDING CHAIR
2    35751C     PURPLE CURRENT CANDLE RING
3    79323G            GOLD  CHERRY LIGHTS
4     21166     COOK WITH WINE METAL SIGN

(986, 2)

In [24]:
len(check_stck['StockCode'].unique())

986

In [25]:
nc.reset_index(inplace=True)

In [26]:
nc.drop(columns='index', inplace=True)

In [27]:
mapper_stck = check_stck.set_index('StockCode').to_dict()['Description']
mapper_stck

{'21350': 'GINGHAM HEART WREATH',
 '84292': 'ROSE PINK METAL FOLDING CHAIR',
 '35751C': 'PURPLE CURRENT CANDLE RING',
 '79323G': 'GOLD  CHERRY LIGHTS',
 '21166': 'COOK WITH WINE METAL SIGN ',
 '21982': 'PACK OF 12 SUKI TISSUES ',
 '84508B': 'STRIPES DESIGN TEDDY',
 '84347': 'ROTATING SILVER ANGELS T-LIGHT HLDR',
 '37446': 'MINI CAKE STAND WITH HANGING CAKES',
 '21133': 'PAINTED BIRD ASSORTED CHRISTMAS',
 '22158': '3 HEARTS HANGING DECORATION RUSTIC',
 '72008': 'FROSTED BOX 9 WHITE T-LIGHT CANDLES',
 '85185F': 'LION SOCK PUPPET',
 '84922': 'PINK BUTTERFLY WASHBAG',
 '20879': 'TREE OF NOAH FESTIVE SCENTED CANDLE',
 '35980C': 'REDWHITE STRIPE SCANDINAVIAN HEART ',
 '22188': 'BLACK HEART CARD HOLDER',
 '22174': 'PHOTO CUBE',
 '72267': 'PREMIUM CHURCH CANDLE',
 '22187': 'GREEN CHRISTMAS TREE CARD HOLDER',
 '16161C': 'WRAP SPEED DEMONS ',
 '47503F': 'ASS FLORAL PRINT TORCH',
 '72045D': 'ROSES ON BLUE TEACUP CANDLE ',
 '84249D': 'FAIRY CAKE GREETINGS CARD',
 '17164B': "ASS COL SMALL SAND GECK

#### Imputing Description

In [28]:
nc['Description'].fillna(0, inplace=True)

In [29]:
nc[nc['Description'] == 0]

Invoice StockCode Description  Quantity         InvoiceDate  Price  \
2989      489659     21350           0       230 2009-12-01 17:39:00    0.0   
3542      489781     84292           0        17 2009-12-02 11:45:00    0.0   
6136      489882    35751C           0        12 2009-12-02 16:22:00    0.0   
6312      489898    79323G           0       954 2009-12-03 09:40:00    0.0   
6336      489903     21166           0        48 2009-12-03 09:57:00    0.0   
...          ...       ...         ...       ...                 ...    ...   
997790    581103     22689           0         4 2011-12-07 11:58:00    0.0   
999337    581203     23406           0        15 2011-12-07 18:31:00    0.0   
999338    581209     21620           0         6 2011-12-07 18:35:00    0.0   
1000963   581234     72817           0        27 2011-12-08 10:33:00    0.0   
1002496   581408     85175           0        20 2011-12-08 14:06:00    0.0   

        Customer ID         Country Day of Week Date Month  Year Day of Year  \
2989          imp26  United Kingdom     Tuesday   01   Dec  2009         335   
3542          imp29  United Kingdom   Wednesday   02   Dec  2009         336   
6136          imp37  United Kingdom   Wednesday   02   Dec  2009         336   
6312          imp39  United Kingdom    Thursday   03   Dec  2009         337   
6336          imp42  United Kingdom    Thursday   03   Dec  2009         337   
...             ...             ...         ...  ...   ...   ...         ...   
997790      imp8084  United Kingdom   Wednesday   07   Dec  2011         341   
999337      imp8092  United Kingdom   Wednesday   07   Dec  2011         341   
999338      imp8098  United Kingdom   Wednesday   07   Dec  2011         341   
1000963     imp8106  United Kingdom    Thursday   08   Dec  2011         342   
1002496     imp8110  United Kingdom    Thursday   08   Dec  2011         342   

        Week of Year Time Hour  Hour  
2989              48  17:39:00    17  
3542              48  11:45:00    11  
6136              48  16:22:00    16  
6312              48  09:40:00     9  
6336              48  09:57:00     9  
...              ...       ...   ...  
997790            49  11:58:00    11  
999337            49  18:31:00    18  
999338            49  18:35:00    18  
1000963           49  10:33:00    10  
1002496           49  14:06:00    14  

[1605 rows x 16 columns]

In [30]:
for i in range(len(nc['Description'])):
    if nc.loc[i, 'Description'] == 0 and nc.loc[i, 'StockCode'] in mapper_stck.keys():
        nc.loc[i, 'Description'] = mapper_stck[nc.loc[i, 'StockCode']]

In [31]:
nc = nc[nc['Description'] != 0]

In [32]:
nc.isna().sum()

Invoice         0
StockCode       0
Description     0
Quantity        0
InvoiceDate     0
Price           0
Customer ID     0
Country         0
Day of Week     0
Date            0
Month           0
Year            0
Day of Year     0
Week of Year    0
Time Hour       0
Hour            0
dtype: int64

In [33]:
print(f" Still got {nc[nc['Price'] == 0].shape[0]} rows where Price are equal to Zero (0). I'm going to drop all price where have a Zero(o) values.")

 Still got 2504 rows where Price are equal to Zero (0). I'm going to drop all price where have a Zero(o) values.


**It just because the business logic where price is zero(0) is impossible, despite of assuming the zero values have another meaning in the real case (should described at features description)**

In [34]:
nc = nc[nc['Price'] != 0]

### g. `Amount` Additional Features

- It's a additional featrues to get insight about *Sales Amount*

In [35]:
nc['Amount'] = nc['Quantity'] * nc['Price']

In [36]:
display(nc.head(2), nc.shape)

Invoice StockCode                          Description  Quantity  \
0   489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1   489434    79323P                   PINK CHERRY LIGHTS        12   

          InvoiceDate  Price Customer ID         Country Day of Week Date  \
0 2009-12-01 07:45:00   6.95     13085.0  United Kingdom     Tuesday   01   
1 2009-12-01 07:45:00   6.75     13085.0  United Kingdom     Tuesday   01   

  Month  Year Day of Year Week of Year Time Hour  Hour  Amount  
0   Dec  2009         335           48  07:45:00     7    83.4  
1   Dec  2009         335           48  07:45:00     7    81.0

(1003214, 17)

**Rearange the `DataFrame`**

In [37]:
nc = nc[['Invoice', 'StockCode', 'Description', 'Quantity', 'Price', 'Amount', 'Customer ID', 'Country', 
         'InvoiceDate', 'Day of Week', 'Date', 'Month', 'Year', 'Day of Year', 'Week of Year', 'Time Hour', 'Hour']]

In [40]:
nc.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1003214 entries, 0 to 1005779
Data columns (total 17 columns):
 #   Column        Non-Null Count    Dtype         
---  ------        --------------    -----         
 0   Invoice       1003214 non-null  int64         
 1   StockCode     1003214 non-null  object        
 2   Description   1003214 non-null  object        
 3   Quantity      1003214 non-null  int64         
 4   Price         1003214 non-null  float64       
 5   Amount        1003214 non-null  float64       
 6   Customer ID   1003214 non-null  object        
 7   Country       1003214 non-null  object        
 8   InvoiceDate   1003214 non-null  datetime64[ns]
 9   Day of Week   1003214 non-null  object        
 10  Date          1003214 non-null  object        
 11  Month         1003214 non-null  object        
 12  Year          1003214 non-null  object        
 13  Day of Year   1003214 non-null  object        
 14  Week of Year  1003214 non-null  object        
 15